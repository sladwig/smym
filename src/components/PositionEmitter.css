
.position-emitter {
    position: relative;
}
.position-area-overlay {
    position: absolute;
    width: 100%;
    height: 100%;
    left: 0;
    top: 0;
    display: inline-flex;
    flex-direction: row;
    cursor: text;
}
.position-area {
    flex: 1 0 0;
}
