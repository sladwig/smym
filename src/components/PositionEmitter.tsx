import React, { ReactNode } from 'react';
import './PositionEmitter.css';
import { setState } from './TokenInput';

interface IPositionEmitterProps {
    position?: number;
    value?: string;
    children?: ReactNode;
    onClick?: (position: number) => () => any;
}
export const PositionEmitter = ({
    children,
    position = 0,
    value = '',
    onClick = updatePosition,
}: IPositionEmitterProps) => {
    return (
        <span className="position-emitter">
            {children ? children : value}
            <span className="position-area-overlay">
                <span className="position-area" onClick={onClick(position)}></span>
                <span className="position-area" onClick={onClick(position + 1)}></span>
            </span>
        </span>
    );
};

const updatePosition = (position: number) => () => setState({ position, external: true });
