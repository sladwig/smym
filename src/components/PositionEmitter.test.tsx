import { render, fireEvent } from '@testing-library/react';
import React from 'react';
import { PositionEmitter } from './PositionEmitter';

describe('PositionEmitter', () => {
    test('renders', () => {
        const { container } = render(<PositionEmitter value="a" />);
        expect(container).toMatchSnapshot();
    });

    test('is clickable and emits click position', () => {
        const mock = jest.fn();
        const onClick = (pos: number) => () => mock(pos);
        const { container } = render(<PositionEmitter value={'a'} onClick={onClick} />);
        container.querySelectorAll('.position-area').forEach(area => {
            fireEvent.click(area);
        });
        expect(mock.mock.calls).toEqual([[0], [1]]);
    });

    test('can set position', () => {
        const mock = jest.fn();
        const onClick = (pos: number) => () => mock(pos);
        const { container } = render(
            <PositionEmitter value={'ab'} onClick={onClick} position={3} />,
        );
        container.querySelectorAll('.position-area').forEach(area => {
            fireEvent.click(area);
        });
        expect(mock.mock.calls).toEqual([[3], [4]]);
    });
});
