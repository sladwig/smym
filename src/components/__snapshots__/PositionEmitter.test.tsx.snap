// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PositionEmitter renders 1`] = `
<div>
  <span
    class="position-emitter"
  >
    a
    <span
      class="position-area-overlay"
    >
      <span
        class="position-area"
      />
      <span
        class="position-area"
      />
    </span>
  </span>
</div>
`;
